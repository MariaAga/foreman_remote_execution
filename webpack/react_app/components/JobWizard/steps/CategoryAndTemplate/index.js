import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import URI from 'urijs';
import { get } from 'foremanReact/redux/API';
import {
  selectJobCategories,
  selectJobTemplates,
  selectJobCategoriesStatus,
} from '../../JobWizardSelectors';
import { CategoryAndTemplate } from './CategoryAndTemplate';

import { JOB_TEMPLATES, JOB_CATEGORIES } from '../../JobWizardConsts';

const ConnectedCategoryAndTemplate = ({
  jobTemplate,
  setJobTemplate,
  category,
  setCategory,
}) => {
  const dispatch = useDispatch();

  const jobCategoriesStatus = useSelector(selectJobCategoriesStatus);
  useEffect(() => {
    if (!jobCategoriesStatus) {
      dispatch(
        get({
          key: JOB_CATEGORIES,
          url: '/ui_job_wizard/categories',
          handleSuccess: response =>
            setCategory(response.data.job_categories[0] || ''),
        })
      );
    }
  }, [jobCategoriesStatus, dispatch, setCategory]);

  const jobCategories = useSelector(selectJobCategories);

  useEffect(() => {
    if (category) {
      const templatesUrl = new URI('/api/v2/job_templates');
      dispatch(
        get({
          key: JOB_TEMPLATES,
          url: templatesUrl.addSearch({
            search: `job_category="${category}"`,
          }),
        })
      );
    }
  }, [category, dispatch]);

  const jobTemplates = useSelector(selectJobTemplates);

  return (
    <CategoryAndTemplate
      jobTemplates={jobTemplates}
      jobCategories={jobCategories}
      setJobTemplate={setJobTemplate}
      selectedTemplateID={jobTemplate}
      setCategory={setCategory}
      selectedCategory={category}
    />
  );
};

ConnectedCategoryAndTemplate.propTypes = {
  jobTemplate: PropTypes.number,
  setJobTemplate: PropTypes.func.isRequired,
  category: PropTypes.string.isRequired,
  setCategory: PropTypes.func.isRequired,
};
ConnectedCategoryAndTemplate.defaultProps = { jobTemplate: null };

export default ConnectedCategoryAndTemplate;
