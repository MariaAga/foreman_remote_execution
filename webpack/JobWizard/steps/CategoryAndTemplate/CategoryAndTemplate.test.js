import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { CategoryAndTemplate } from './CategoryAndTemplate';

const baseProps = {
  setJobTemplate: jest.fn(),
  selectedTemplateID: 190,
  setCategory: jest.fn(),
};
const fixtures = {
  'renders with props': {
    ...baseProps,
    jobCategories: [
      'Commands',
      'Ansible Playbook',
      'Ansible Galaxy',
      'Ansible Roles Installation',
    ],
    jobTemplates: [
      {
        id: 190,
        name: 'ab Run Command - SSH Default clone',
        job_category: 'Commands',
        provider_type: 'SSH',
        snippet: false,
      },
      {
        id: 168,
        name: 'Ansible Roles - Ansible Default',
        job_category: 'Ansible Playbook',
        provider_type: 'Ansible',
        snippet: false,
      },
      {
        id: 170,
        name: 'Ansible Roles - Install from git',
        job_category: 'Ansible Roles Installation',
        provider_type: 'Ansible',
        snippet: false,
      },
    ],
    selectedCategory: 'I am a category',
  },
  'render with error': {
    ...baseProps,
    errors: { allTemplatesError: 'I have an error' },
  },
};

describe('CategoryAndTemplate', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(CategoryAndTemplate, fixtures));
});
