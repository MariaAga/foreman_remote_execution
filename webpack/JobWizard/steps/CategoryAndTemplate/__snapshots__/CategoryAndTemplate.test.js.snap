// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryAndTemplate rendering render with error 1`] = `
<Fragment>
  <Title
    headingLevel="h2"
  >
    Category And Template
  </Title>
  <Text
    component="p"
  >
    All fields are required.
  </Text>
  <Form>
    <SelectField
      fieldId="job_category"
      isDisabled={false}
      label="Job category"
      options={Array []}
      placeholderText=""
      setValue={[Function]}
      value={null}
    />
    <GroupedSelectField
      fieldId="job_template"
      groups={Array []}
      isDisabled={true}
      label="Job template"
      placeholderText="Error"
      selected={null}
      setSelected={[MockFunction]}
    />
    <Alert
      title="Errors:"
      variant="danger"
    >
      <span>
        Templates list failed with:
         
        I have an error
      </span>
    </Alert>
  </Form>
</Fragment>
`;

exports[`CategoryAndTemplate rendering renders with props 1`] = `
<Fragment>
  <Title
    headingLevel="h2"
  >
    Category And Template
  </Title>
  <Text
    component="p"
  >
    All fields are required.
  </Text>
  <Form>
    <SelectField
      fieldId="job_category"
      isDisabled={false}
      label="Job category"
      options={
        Array [
          "Commands",
          "Ansible Playbook",
          "Ansible Galaxy",
          "Ansible Roles Installation",
        ]
      }
      placeholderText=""
      setValue={[Function]}
      value="I am a category"
    />
    <GroupedSelectField
      fieldId="job_template"
      groups={
        Array [
          Object {
            "groupLabel": "SSH",
            "options": Array [
              Object {
                "label": "ab Run Command - SSH Default clone",
                "value": 190,
              },
            ],
          },
          Object {
            "groupLabel": "Ansible",
            "options": Array [
              Object {
                "label": "Ansible Roles - Ansible Default",
                "value": 168,
              },
              Object {
                "label": "Ansible Roles - Install from git",
                "value": 170,
              },
            ],
          },
        ]
      }
      isDisabled={false}
      label="Job template"
      placeholderText=""
      selected="ab Run Command - SSH Default clone"
      setSelected={[MockFunction]}
    />
  </Form>
</Fragment>
`;
